### PR TITLE
Automated cherry pick of #101047: Only use dualstack if the node and config supports it

### PR DESCRIPTION
**Pull request by jsturtevant**
Cherry pick of #101047 on release-1.21.

#101047: Only use dualstack if the node and config supports it

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
/priority important-soon
/triage accepted
/kind bug

**Comment by neolit123** [comment]
LGTM, based on the review of https://github.com/kubernetes/kubernetes/pull/101047
needs lgtm/approve for the proxy packages.

**Comment by jsturtevant** [comment]
/assign @caseydavenport 
for kube-proxy

**Comment by jsturtevant** [comment]
/assign @kubernetes/release-managers 
ready for release managers approval

File: cmd/kube-proxy/app/server.go
Comment (khenidak) [inline cmd/kube-proxy/app/server.go:827]: There is a util pkg under proxy where there is a function that returns all addresses for a node as `sets.String{}` -- Why are we using a different approach here?

Comment (aojea) [reply cmd/kube-proxy/app/server.go:827]: we don't need all addresses of a node, we need to have a consistent way to detect the primary ip of the node to deal with the kube-proxy nodeIP chaos, related to https://github.com/kubernetes/kubernetes/pull/104691/files#r700086018
and explained https://github.com/kubernetes/kubernetes/issues/95515

**Comment by marosset** [comment]
@puerco @cpanato can one of you take a look here?
Thanks!

**Review by cpanato — APPROVED** [review]
/lgtm

**Pull request merged**
